### ISSUE TITLE
Feature Suggestion: show password toggle on password fields 

### ISSUE THREAD

**Issue by anthony-hull** [Expensify/App#2734]
View password feature on password fields
## Motivation
Improve accessibility for users.
## Implementation
I think it would be good to replicate the UX of Edge browser where when you type it shows an icon button inside the field to toggle show password. 
You should also be able to tab to the icon on desktop to maintain keyboard nav. 

Edge cases:
Hiding native view password icon in Edge
Password managers tamper with the DOM in this location

Upwork job: https://www.upwork.com/jobs/~01f52f49fe3707fca2

**Comment by mallenexpensify** [comment]
Holding until https://github.com/Expensify/Expensify.cash/issues/1517 is closed then I'll start a review.
@anthony-hull I'm going to assign to you for now, can you kick it back to me once 1517 is closed and I'll take it from there?

**Comment by rdjuric** [comment]
Was about to suggest this feature and found an open issue for it 😃

cc'ing @mallenexpensify since the PR you mentioned is closed now.

**Comment by anthony-hull** [comment]
Ahh yes. I forgot this one was open :)
Thank you Rafael

**Comment by mallenexpensify** [comment]
@anthony-hull can you provide more detail on what you're recommending, and possibly include a screenshot or vid? Any chance this has been discussed in Slack, #expensify-open-source yet? If so, can you drop in a link?  Thanks

**Comment by anthony-hull** [comment]
Yes we had a [discussion on slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1620306620454000)
I will follow up with a video on Monday. I'm busy the next 4 days.

**Comment by anthony-hull** [comment]
I think it's best to show the password by default on mobile devices. You can trivially move your screen in a way to obscure the password as you type. Additionally as you type there is a large on screen keyboard that people can watch you enter it on. I'm not sure there is much of a security gain from it vs the loss of UX.

On desktop, mask passwords by default, but add a “Show” option so that people can remember the password and check their work.
On mobile, unmask passwords by default, but add a “Hide” option. It’s especially hard to type a complex, secure password on mobile.
Unmasking passwords on mobile is not really hurting security very much. Mobile keyboards indicate exactly which letters were typed, anyway.

Ironically now I think of it, I just added the confirm password field in the set password form. If you allow the user to check their work or show by default you can remove the need for repetition. We could have just one entry box and make it easier.

I think behaviour of showing and hiding that mailchip implemented is a nice way to go:
https://login.mailchimp.com/signup/
![image](https://user-images.githubusercontent.com/47184118/123687728-2d59bc80-d849-11eb-930c-2df3b410ba9f.png)
![image](https://user-images.githubusercontent.com/47184118/123687768-36e32480-d849-11eb-9c5e-0990a78733b2.png)


Or how Edge does it with the icon inside the input:
![image](https://user-images.githubusercontent.com/47184118/123688043-7f024700-d849-11eb-8b64-8cc368eba368.png)
![image](https://user-images.githubusercontent.com/47184118/123688081-87f31880-d849-11eb-865a-7a3b98a9841a.png)

We could hide the password again when the box loses focus.

**Comment by mallenexpensify** [comment]
Thanks @anthony-hull , I'm double checking to make sure this isn't something that's already being worked on internally.  I'll followup tomorrow

**Comment by mallenexpensify** [comment]
Thanks for holding, nobody mentioned this is something that's currently be worked on. I agree this is feature that would be useful. First, can you take a peek at this issue to see if any work being done there would help on mobile for this issue?  https://github.com/Expensify/Expensify.cash/issues/3108 (I have no idea, just searched 'show password' in the repo to ensure something wasn't already being worked on). 

Then, can you submit a proposal for how you want to implement this and I'll assign an engineer to review? 
Thanks

**Comment by anthony-hull** [comment]
It would make sense to use the new input component in that issue. I'm happy to hold until it's merged.
When `secureTextEntry` is passed into the component I would change the behaviour to: 

Implement the show/hide icon as a background image svg with onClick handlers to change the visibility of the text input. 
The visibility can be controlled by changing the props passed to the wrapped text input field.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @deetergp (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mallenexpensify** [comment]
I used `Exported` to assign an engineer for review. @deetergp , can you review @anthony-hull 's proposal? 

@puneetlath , I unassigned myself then added `External` because I'm OOO next week.  Also, I removed `Exported` label because it **hasn't** been posted to Upwork yet, it makes more sense to me to get Deeter to agree on the solution before posting.

**Comment by deetergp** [comment]
The feature @anthony-hull proposes sounds good, but we're going to need a more detailed description about how he is going to do it before moving forward.

**Comment by anthony-hull** [comment]
The exact UX isn't agreed. 
https://github.com/Expensify/Expensify.cash/pull/3414/files

The following props are set for TextInputs are used for passwords:
```
secureTextEntry
autoCompleteType="password"
textContentType="password"
```
When `textContentType` is set to `password` in the props of ExpensiTextInput:
src/components/ExpensiTextInput/index.js
I will add an SVG of an eye or whatever icon I'm given to the input. Above or inside the input depending on the designs. 
I will hold state in the ExpensiTextInput component for show password or not. When you click on the icon it will toggle that state. 
The state will control the Icon shown and if the secureTextEntry is true or false when passed into the wrapped TextInput component:

<Icon name={this.state.iconType} onPress={() => this.toggleShow />
<Input secureTextEntry={this.state.show}  />

**Comment by deetergp** [comment]
Your proposal makes sense and sounds good, @anthony-hull. Thanks! 

@puneetlath I am unsure of the next step, since this has not been posted on Upwork.

**Comment by puneetlath** [comment]
Created the Upwork job and invited @anthony-hull to it: https://www.upwork.com/jobs/~01f52f49fe3707fca2

**Comment by anthony-hull** [comment]
I accepted the job on UW. I can start work with what I have. But in parallel:
Ideally I would like an eye icon of two states provided by your designers.
I see you have: 
![image](https://user-images.githubusercontent.com/47184118/124664661-b3a67c00-dea3-11eb-9d0b-455cb01eb3eb.png) in Expensify.cash\assets\images

But I think we need a toggled state with the eye closed to show the different hidden/shown states.

**Comment by puneetlath** [comment]
@Expensify/design anyone able to help with icons for this?

**Comment by shawnborton** [comment]
I can grab this quick as we already have the icon, I just need to export it to an SVG:
![image](https://user-images.githubusercontent.com/2319350/124676575-a2de1200-de73-11eb-941a-d808949eabab.png)

**Comment by shawnborton** [comment]
Here is the svg: [eye-disabled.svg.zip](https://github.com/Expensify/Expensify.cash/files/6773210/eye-disabled.svg.zip)

**Comment by anthony-hull** [comment]
I'm still waiting on https://github.com/Expensify/Expensify.cash/pull/3414 to be merged for this issue as I will be working in that new component.
I've started some work on it, however.

**Comment by deetergp** [comment]
Thanks for the update @anthony-hull.

**Comment by deetergp** [comment]
It looks like https://github.com/Expensify/App/pull/3414 is getting close. We've got 1 out of 3 reviewers' approval on the PR.

**Comment by anthony-hull** [comment]
Now the PR has merged I will re-start work on this tomorrow.

**Comment by deetergp** [comment]
Thanks for the update, @anthony-hull 👍

**Comment by anthony-hull** [comment]
Do we want to be able to control the visibility of all password fields on a page together?

https://user-images.githubusercontent.com/47184118/129714985-754dafe5-2a77-4a23-9d79-142887aef18a.mp4

in this video you can see they are toggling separately as the state is within the input.
Please note I intend on removing the second new password input field to improve the UX now we can check the password via making it visible.

**Comment by shawnborton** [comment]
I think it makes sense if each icon only controlled the input that it was connected to. Otherwise we should think of a different kind of UI control to toggle the visibility for all fields on a page.

**Comment by puneetlath** [comment]
@anthony-hull just curious how this is going.

**Comment by anthony-hull** [comment]
I have a question in the draft PR, I started a review. I'll make it a full PR as maybe that is what is confusing my intentions.

**Comment by deetergp** [comment]
There's some discussion going on in the draft PR, but this is continuing to move forward.

**Comment by puneetlath** [comment]
Hi @anthony-hull. Just checking in again to see how this is going 😄.

**Comment by anthony-hull** [comment]
Yeah sorry I'm on holiday until Monday :)
I have the code ready just still need to test on Android and Desktop

**Comment by deetergp** [comment]
Heya, @anthony-hull, have you made any progress with this issue?

**Comment by anthony-hull** [comment]
Yes I'm waiting for a PR review.
I messaged the assigned staff member yesterday to let them know it's ready.

**Comment by johnmlee101** [comment]
Do we think that this should also be on hold due to `N6`?

**Comment by puneetlath** [comment]
Yes, I think we should hold merging it until n6 is released. We can still approve it, but I think we should hold merging. And we'll provide @anthony-hull with the bonus as compensation for the delay.

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by deetergp** [comment]
Keep your shirt on, Melvin — it's still a weekly.

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @deetergp, @puneetlath, @anthony-hull eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by puneetlath** [comment]
@anthony-hull @johnmlee101 how's this one going?

**Comment by anthony-hull** [comment]
Sorry notification got missed by me. PR is ready hopefully for final review. Just need to test on all platforms again

**Comment by puneetlath** [comment]
Paid with n6 bonus. Thanks!
